### PR TITLE
Code verification with LiquidHaskell

### PR DESCRIPTION
**Pull request by CodiePP**
description
-----------

- [x] all code now validates with LiquidHaskell

checklist
---------

- [x] compiles (`cabal new-clean; cabal new-build`)
- [x] tests run successfully (`cabal new-test`)
- [ ] documentation added and created (`cd docs; nix-shell --run make`)
- [ ] link to an issue
- [ ] link to an epic
- [ ] add estimate points
- [ ] add milestone (the same as the linked issue)

### CODE REVIEW

**Comment by CodiePP** [comment]
* we need a script that runs _LH_ over all *.lhs files in `src/`
* can this be added to `CI`?

File: iohk-monitoring/src/Cardano/BM/Data/Aggregated.lhs [hunk 1]
```diff
@@ -118,42 +119,50 @@ instance Num Measurable where
     (+) (Bytes a)        (Bytes b)        = Bytes        (a+b)
     (+) (PureI a)        (PureI b)        = PureI        (a+b)
     (+) (PureD a)        (PureD b)        = PureD        (a+b)
-    (+)  _                _               = error "Trying to add values with different units"
+    (+) a                _                = a
```
Comment (365andreas) [inline iohk-monitoring/src/Cardano/BM/Data/Aggregated.lhs]: `fail` could warn as that we are wrongly adding different values.

File: iohk-monitoring/src/Cardano/BM/Output/Switchboard.lhs [hunk 2]
```diff
@@ -286,12 +274,16 @@ setupBackends :: ToObject a
...
+  where
+    setupBackendsAcc [] acc = return acc
+    setupBackendsAcc (bk : r) acc = do
+        setupBackend' bk c sb >>= \case
+            Nothing -> setupBackendsAcc r acc
+            Just be -> setupBackendsAcc r ((bk,be) : acc)
 
 setupBackend' :: ToObject a => BackendKind -> Configuration -> Switchboard a -> IO (Maybe (Backend a))
-setupBackend' SwitchboardBK _ _ = error "cannot instantiate a further Switchboard"
+setupBackend' SwitchboardBK _ _ = return Nothing -- error "cannot instantiate a further Switchboard"
```
[the remark this answers is not in the document]
Comment (CodiePP) [reply iohk-monitoring/src/Cardano/BM/Output/Switchboard.lhs]: yes,

File: iohk-monitoring/src/Cardano/BM/Data/MonitoringEval.lhs [hunk 3]
```diff
@@ -91,11 +92,11 @@ instance Eq MEvExpr where
     (==) _                _                = False
 
 instance FromJSON MEvExpr where
-    parseJSON (String s) =
+    parseJSON o@(String s) =
         case parseEither s of
-            Left e     -> error e
+            Left _e    -> typeMismatch "String" o
```
[the remark this answers is not in the document]
Comment (CodiePP) [reply iohk-monitoring/src/Cardano/BM/Data/MonitoringEval.lhs]: yes

File: iohk-monitoring/src/Cardano/BM/Output/Switchboard.lhs (hunk 2 above)
[the remark this answers is not in the document]
Comment (CodiePP) [reply iohk-monitoring/src/Cardano/BM/Output/Switchboard.lhs]: redone

**Comment by CodiePP** [comment]
rebased, ready for review

File: iohk-monitoring/src/Cardano/BM/Output/Aggregation.lhs [hunk 4]
```diff
@@ -263,50 +293,62 @@ We use Welford's online algorithm to update the estimation of mean and variance
...
+                                         , fold = mkTimestamp
+                                         , fbasic = updateBaseStats 1 v (fbasic s)
+                                         , fdelta = updateBaseStats 2 deltav (fdelta s)
+                                         , ftimed = updateBaseStats 2 timediff (ftimed s)
+                                         }
   where
+    deltav = subtractMeasurable v (flast s)
     mkTimestamp = utc2ns (tstamp lme)
+    timediff = Nanoseconds $ fromInteger $ (getInteger mkTimestamp) - (getInteger $ fold s)
     utc2ns (UTCTime days secs) =
```
Comment (365andreas) [inline iohk-monitoring/src/Cardano/BM/Output/Aggregation.lhs]: `utc2ns` is also defined here:
https://github.com/input-output-hk/iohk-monitoring-framework/pull/303/files#diff-cbc1ae38e6be73f27a17b5791e7cb91aR180
Should we put it into another file in order to both use it from there?

Comment (CodiePP) [reply iohk-monitoring/src/Cardano/BM/Output/Aggregation.lhs]: yes, this could be done if we find a good place for it.

**Pull request merged**